### PR TITLE
Memoize cuTT/hipTT plans again

### PR DESCRIPTION
**Pull request by benson31**
#2272 was hastily implemented as "working correctly" was more important than "working well" at the time. Profiling discovered some (not surprising) issues with constructing plans each iteration (again, not surprising). This restores the memoization but takes care to consider the forward and inverse actions separately. It does NOT optimize further to the case in which a permutation is its own inverse; in this case, at least two plans will still be constructed and stored.

### CODE REVIEW

**Review by tbennun — CHANGES_REQUESTED** [review]

File: src/layers/transform/permute/cutt_permuteimpl.hpp [hunk 1]
```diff
@@ -183,6 +194,12 @@ class cuTT_PermuteImpl
   ColMajorPerm m_inv_perm;
   DimsType m_input_dims;
   DimsType m_output_dims;
+
+  // Plan memoization -- lazily constructed.
+  mutable PlanMap m_fwd_plans;
+  mutable PlanMap m_inv_plans;
+  mutable Plan m_sample_fwd_plan = 0;
+  mutable Plan m_sample_inv_plan = 0;
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/layers/transform/permute/cutt_permuteimpl.hpp:202]: Nope. [unsigned int](https://github.com/DmitryLyakh/hipTT/blob/master/src/cutt.h#L31) But I will make the literal `0U`.

[the remark this answers is not in the document]
Comment (tbennun) [reply src/layers/transform/permute/cutt_permuteimpl.hpp:202]: Thanks

File: src/layers/transform/permute/cutt_permuteimpl.hpp [hunk 2]
```diff
@@ -288,58 +334,59 @@ void cuTT_PermuteImpl::inverse_permute(
   if (in.Width() == El::Int{0})
     return;
 
-  // Aliases to help this read better.
-  auto const& in_dims = m_output_dims;
-  auto const& out_dims = m_input_dims;
-  auto const& perm = m_inv_perm;
-
-  auto multisync =
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/layers/transform/permute/cutt_permuteimpl.hpp:296]: Fixed. I intended to move down to `do_*_permute`, but it was missed. Thanks for the catch!

File: src/layers/transform/permute/cutt_permuteimpl.hpp [hunk 3]
```diff
@@ -148,9 +148,14 @@ class cuTT_PermuteImpl
   void swap(cuTT_PermuteImpl& other);
   ///@}
 
+private:
+  using Plan = cuttHandle;
+  using PlanMap = std::unordered_map<El::Int, Plan>;
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/layers/transform/permute/cutt_permuteimpl.hpp]: Done.

**Review by tbennun — APPROVED** [review]

**Pull request merged**